app/main.py

```diff
@@ -0,0 +1,124 @@
+# FastAPI App
+# ############
+LOGGER = logging.getLogger("api")
+
+APP = fastapi.FastAPI(
+    title="Coronavirus Tracker",
+    description="API for tracking the global coronavirus (COVID-19, SARS-CoV-2) outbreak.",
+    version="2.1.0",
+    prefix="/v2-1",
+    docs_url="/v2-1",
+    redoc_url="/docs",
+)
+
+# #####################
+# Middleware
+#######################
+
+
+# TODO this could probably just be a FastAPI dependency
+@APP.middleware("http")
+async def add_datasource(request: fastapi.Request, call_next):
+    """Attach the data source to the request.state."""
+    source = request.query_params.get("source", default="jhu")
+    request.state.source = data_source(source)
+    LOGGER.info(f"source: {request.state.source.__class__.__name__}")
+    response = await call_next(request)
+    return response
+
+
+# ################
+# Exception Handler
+# ################
+
+
+@APP.exception_handler(pydantic.error_wrappers.ValidationError)
+async def handle_validation_error(
+    request: fastapi.Request, exc: pydantic.error_wrappers.ValidationError
+):
+    return fastapi.responses.JSONResponse({"message": exc.errors()}, status_code=422)
+
+
+# ################
+# Routes
+# ################
+
+
+@APP.get("/latest", response_model=models.Latest)
+def get_latest(request: fastapi.Request):
+    """Getting latest amount of total confirmed cases, deaths, and recoveries."""
+    locations = request.state.source.get_all()
+    return {
+        "latest": {
+            "confirmed": sum(map(lambda location: location.confirmed, locations)),
+            "deaths": sum(map(lambda location: location.deaths, locations)),
+            "recovered": sum(map(lambda location: location.recovered, locations)),
+        }
+    }
+
+
+@APP.get("/locations", response_model=models.AllLocations)
+def get_all_locations(
+    request: fastapi.Request, country_code: str = None, timelines: int = 0
+):
+    # Retrieve all the locations.
+    locations = request.state.source.get_all()
+
+    # Filtering my country code if provided.
+    if country_code:
+        locations = list(
+            filter(
+                lambda location: location.country_code == country_code.upper(),
+                locations,
+            )
+        )
+    # FIXME: timelines are not showing up
+    return {
+        "latest": {
+            "confirmed": sum(map(lambda location: location.confirmed, locations)),
+            "deaths": sum(map(lambda location: location.deaths, locations)),
+            "recovered": sum(map(lambda location: location.recovered, locations)),
+        },
+        "locations": [location.serialize(timelines) for location in locations],
+    }
+
+
+@APP.get("/locations/{id}", response_model=models.Location)
+def get_location_by_id(request: fastapi.Request, id: int, timelines: int = 1):
+    return {"location": request.state.source.get(id).serialize(timelines)}
+
+
+# mount the existing Flask app to /v2
+APP.mount("/v2", WSGIMiddleware(create_app()))
```

Comment:
I'm not sure how this behaves with the `/v1` endpoints. At least locally, the `/v2` endpoints appear to be functional, including the redirect to your github page when hitting `/v2`.

Reply:
So FastAPI works with flask or? The v1 endpoints can be found at:
1. https://coronavirus-tracker-api.herokuapp.com/all
2. https://coronavirus-tracker-api.herokuapp.com/confirmed
3. https://coronavirus-tracker-api.herokuapp.com/deaths
4. https://coronavirus-tracker-api.herokuapp.com/recovered

I have checked the logs, and they are still in use by a lot of apps, so they need to work.

Reply:
@ExpDev07 
So because /v1 and /v2 are defined into the same app, this appears to put both `/v2` and `/v1` behind the `/v2` prefix.

In order to separate them I believe I need to create 2 separate flask apps and mount them separately.

For example `/confirmed` is part of `/v1`, but I can access it locally at `/v2/confirmed` when it should be `/v1/confirmed`.

Reply:
I'll probably need to do something like...
```python
APP.mount("/v1", WSGIMiddleware(create_app("v1")))
APP.mount("/v2", WSGIMiddleware(create_app("v2")))
```

Reply:
It should be ``/confirmed`` without any prefix :). Hmm, two different flask apps? FastAPI doesn't support blueprinting?

Reply:
@Kilo59 it looks like it supports different routers. Look here: https://www.errietta.me/blog/python-fastapi-intro/, and search for ``include_router``.

Reply:
> So FastAPI works with flask or?

@ExpDev07 
Yes, it does.
https://fastapi.tiangolo.com/advanced/wsgi/
My thinking is that we could start deploying with FastAPI with versions 1 and 2 ["mounted"](https://fastapi.tiangolo.com/advanced/wsgi/).
 
And with a FastAPI version of `/v2` that lives at `/v2-1` (or `/v2-beta`), once we have confirmed that it works under stress and smoothed out any edge-cases, we could remove the flask mounted v2 (and keep mounting v1) and move the `FastAPI` version to `/v2`.

Hopefully, that made sense 😄

Reply:
> @Kilo59 it looks like it supports different routers. Look here: https://www.errietta.me/blog/python-fastapi-intro/, and search for `include_router`.
> 
> So we'll just have a different `APIRouter` per api version. Something like:
> Seems to be exactly what we're looking for.

Yes, we can do that too, but then we'd have to rewrite those versions with FastAPI instead of just mounting them and forgetting about it.
Which we definitely could do, as it would mean they could have higher throughput and their own swagger documentation.

Reply:
I mean if the old v1 enpoints work then, then I guess it's fine.

Reply:
I'd be happy to get on a screen share, discord etc. to talk about this stuff (and deployment configuration, etc.).

Reply:
My Discord # is: Marius Big Ounce#2997.

Reply:
Just sent you an invite.
Kilo59#8819

Reply:
I think I accidentally declined it, sent you one now.